Adding initial EndpointSlice documentation (#16018)
kubernetes/website: content/en/docs/concepts/services-networking/endpointslice.md

---
reviewers:
- freehan
title: Endpoint Slice
feature:
  title: Endpoint Slices
  description: >
    Scalable tracking of network endpoints in a Kubernetes cluster.

content_template: templates/concept
weight: 10
---

{{% capture overview %}}

{{< feature-state for_k8s_version="v1.16" state="alpha" >}}
{{< glossary_definition term_id="endpointslice" length="short" >}}

_Endpoint Slices_ provide a simple way to track network endpoints within a
Kubernetes cluster. They offer a more scalable alternative to Endpoints
resources within Kubernetes.

{{% /capture %}}

{{% capture body %}}

> ## Motivation

Reviewer: Overall, in concept-type documentation, I am wary of anchoring a page at a particular point in time. This section would be great for a blog post; in the concepts section, it might age.

I can imagine that this page look like this in many months' time, even after Endpoint Slices have become the norm for Service resources in Kubernetes.
I recommend keeping this section trim and moving it to the foot of the page.

If needed (to avoid a holdup), that could be something to tweak after the v1.16 release.

Author: That makes sense. I've shrunk this a bit and moved it to the bottom.

---
reviewers:
- freehan
title: Endpoint Slice
feature:
  title: Endpoint Slices
  description: >
    Scalable tracking of network endpoints in a Kubernetes cluster.

content_template: templates/concept
weight: 10
---

{{% capture overview %}}

> {{< feature-state for_k8s_version="v1.16" state="alpha" >}}

Reviewer 2: @robscott , is the Endpoint Slices feature in the feature gates reference,
https://deploy-preview-16018--kubernetes-io-vnext-staging.netlify.com/docs/reference/command-line-tools-reference/feature-gates/

Author: Missed that initially, thanks for catching that! Should be added now.

---
reviewers:
- freehan
title: Endpoint Slice
feature:
  title: Endpoint Slices
  description: >
    Scalable tracking of network endpoints in a Kubernetes cluster.

content_template: templates/concept
weight: 10
---

{{% capture overview %}}

{{< glossary_definition term_id="endpointslice" length="short" >}}

> {{< feature-state for_k8s_version="v1.16" state="alpha" >}}

Author: Updated, thanks!

> Endpoint Slices provide a simple way to track network endpoints within a

Author: Done, thanks!

---
reviewers:
- freehan
title: Endpoint Slice
feature:
  title: Endpoint Slices
  description: >
    Scalable tracking of network endpoints in a Kubernetes cluster.

content_template: templates/concept
weight: 10
---

{{% capture overview %}}

{{< feature-state for_k8s_version="v1.16" state="alpha" >}}
{{< glossary_definition term_id="endpointslice" length="short" >}}

_Endpoint Slices_ provide a simple way to track network endpoints within a
Kubernetes cluster. They offer a more scalable alternative to Endpoints
resources within Kubernetes.

{{% /capture %}}

{{% capture body %}}

## Endpoint Slice resources {#endpointslice-resource}

In Kubernetes, an Endpoint Slice contains references to a set of network
endpoints. The EndpointSlice controller automatically creates Endpoint Slices

> for a Kubernetes Service with a selector is specified. These Endpoint Slices

Reviewer 2: nit: line 21, `Endpoints resources` vs `Endpoint resources` vs. `Endpoints`. You can probably drop resource 
nit: for a Kubernetes Service when a selector ... ?

Author: Thanks! All this terminology is confusing. Trying to differentiate Kubernetes Endpoints resources from the generic concept of network endpoints is hard to balance, hopefully it's all reasonably clear now, but let me know if there's a better way to communicate the difference here.

will include references to any Pods that match the Service selector. Endpoint
Slices group network endpoints together by unique Service and Port combinations.

As an example, here's a sample EndpointSlice resource for the `example`

> Kubernetes service.

Reviewer 2: nit: Service?

Author: Thanks!

more than 100 endpoints each. Below this scale, Endpoint Slices should map 1:1
with Endpoints and Services and have similar performance.

Endpoint Slices can act as the source of truth for kube-proxy when it comes to
how to route internal traffic. When enabled, they should provide a performance
improvement for services with large numbers of endpoints.

## Motivation

The Endpoints API has provided a simple and straightforward way of
tracking network endpoints in Kubernetes. Unfortunately as Kubernetes clusters
and Services have gotten larger, limitations of that API became more visible.
Most notably, those included challenges with scaling to larger numbers of
network endpoints.

Since all network endpoints for a Service were stored in a single Endpoints
resource, those resources could get quite large. That affected the performance
of Kubernetes components (notably the master control plane) and resulted in
significant amounts of network traffic and processing when Endpoints changed.
Endpoint Slices help you mitigate those issues as well as provide an extensible
platform for additional features such as topological routing.

> ## What's Next

Reviewer 2: nit: you can remove the `What's Next` heading. The shortcode on line 90 should add this heading automatically.

Author: Done, thanks!

---
reviewers:
- freehan
title: Endpoint Slice
feature:
  title: Endpoint Slices
  description: >
    Scalable tracking of network endpoints in a Kubernetes cluster.

content_template: templates/concept
weight: 10
---

{{% capture overview %}}

{{< glossary_definition term_id="endpointslice" length="short" >}}

Endpoint Slices provide a simple way to track network endpoints within a

> Kubernetes cluster. Currently in alpha, they are intended to eventually replace

Reviewer: As per the style guide, [avoid making statements about the future](https://kubernetes.io/docs/contribute/style/style-guide/#avoid-statements-about-the-future)

Author: Fixed, thanks!

the current Endpoints resource within Kubernetes.

{{% /capture %}}

{{% capture body %}}

## Motivation

The current Endpoints API has provided a simple and straightforward way of
tracking network endpoints in Kubernetes. Unfortunately as Kubernetes clusters
and Services have gotten larger, limitations of that API became more visible.
Most notably, those included challenges with scaling to larger numbers of
network endpoints. 

Since all network endpoints for a Service were stored in a single Endpoints
resource, those resources could get quite large. That affected the performance

> of Kubernetes components like kube-proxy and resulted in significant amounts of

Reviewer 3: kube-proxy isn't actually the main victim of it - it's more like master-control-plane (etcd, apiserver)

Author: Updated, thanks!

network traffic when Endpoints changed. With kube-proxy running on each node,
each Endpoints change resulted in the full resource being transmitted to each
node. In some cases, Endpoints resources became large enough to hit the upper
limit on the size of a single object in etcd.

> Endpoint Slices were designed to mitigate those issues as well as provide an

Reviewer: ```suggestion
Endpoint Slices help you mitigate those issues as well as provide an
```

Author: updated, thanks!

extensible platform for additional features such as topological routing.

## Endpoint Slice resources {#endpointslice-resource}

In Kubernetes, an Endpoint Slice contains references to a set of network

> endpoints. Endpoint Slices are automatically created for a Kubernetes Service

Reviewer: > Endpoint Slices are automatically created for a Kubernetes Service
What automatically creates these (eg: the control plane)?

Author: Added more detail here, thanks!

with a selector specified. These Endpoint Slices will include references to any
Pods that match the Service selector. Endpoint Slices group network endpoints 
together by unique Service and Port combinations. By default, Endpoint Slices
will have no more than 100 endpoints each. 

As an example, here's a sample EndpointSlice resource for the `example`
Kubernetes service.

> 

Reviewer 3: I would mention explicitly, that if your service is small enough (by default <=100), the although the API Endpoints vs EndpointSlice differs, conceptually it's the same (1-1 with Service). The real conceptual difference starts for larger services where we no longer have 1-1 correspondence.

Author: Good call, added a reference to this.

```yaml
apiVersion: discovery/v1alpha
kind: EndpointSlice
metadata:
  name: example.abc
  labels:
    kubernetes.io/service-name: example
addressType: IP
ports:
  - name: http
    protocol: TCP
    port: 80
endpoints:
  - addresses:
    - "10.1.2.3"

>     - "2001:db8::1234:5678"

Reviewer: :ok_hand: Nice use of an RFC 3849 address!

    conditions:
      ready: true
    hostname: pod-1
    topology:
      kubernetes.io/hostname: node-1
      topology.kubernetes.io/zone: us-west2-a
```

> ## Enabling Endpoint Slices

Reviewer: Consider moving this to a new Task page, leaving the Concept page short and readable.

Author: I moved this to a new task page, but I'm thinking that task page needs some work, seems fairly light/empty right now.

TODO(robscott)

> ## Future work

Reviewer: As per the style guide, [avoid making statements about the future](https://kubernetes.io/docs/contribute/style/style-guide/#avoid-statements-about-the-future).

Author: Done, thanks!
